### PR TITLE
Scheduler - onAppointmentContextMenu is not triggered for an appointment in the appointment collector (T1181442)

### PR DESCRIPTION
**Pull request by markallenramirez**

### CODE REVIEW

File: packages/devextreme/js/__internal/scheduler/tooltip_strategies/m_desktop_tooltip_strategy.ts [hunk 1]
```diff
@@ -55,4 +56,18 @@ export class DesktopTooltipStrategy extends TooltipStrategyBase {
...
   }
+
+  _onListItemContextMenu(e) {
+    const config = {
+      itemData: e.itemData.appointment,
+      itemElement: e.itemElement,
+      targetedAppointment: e.itemData.targetedAppointment,
+    };
+
+    const createContextMenuEvent = extendFromObject(this._options.fire('mapAppointmentFields', config), e, false);
```
Comment (andrewmakarov) [inline packages/devextreme/js/__internal/scheduler/tooltip_strategies/m_desktop_tooltip_strategy.ts]: why not use the spread operator?

File: packages/devextreme/js/__internal/scheduler/tooltip_strategies/m_desktop_tooltip_strategy.ts [hunk 2]
```diff
@@ -55,4 +56,18 @@ export class DesktopTooltipStrategy extends TooltipStrategyBase {
...
+    delete createContextMenuEvent.itemData;
```
Comment (andrewmakarov) [inline packages/devextreme/js/__internal/scheduler/tooltip_strategies/m_desktop_tooltip_strategy.ts]: Something is not good code, I think it needs to be re-work

**Review by andrewmakarov — CHANGES_REQUESTED** [review]

File: packages/devextreme/js/__internal/scheduler/tooltip_strategies/m_desktop_tooltip_strategy.ts (hunk 2 above)
Comment (markallenramirez) [reply packages/devextreme/js/__internal/scheduler/tooltip_strategies/m_desktop_tooltip_strategy.ts]: reworked as discussed

File: packages/devextreme/js/__internal/scheduler/tooltip_strategies/m_desktop_tooltip_strategy.ts (hunk 1 above)
Comment (markallenramirez) [reply packages/devextreme/js/__internal/scheduler/tooltip_strategies/m_desktop_tooltip_strategy.ts]: reworked as discussed

File: packages/devextreme/testing/tests/DevExpress.ui.widgets.scheduler/common.events.tests.js [hunk 3]
```diff
@@ -718,4 +718,44 @@ QUnit.module('Events', {
...
+            currentDate: new Date(2023, 7, 18),
+            onAppointmentContextMenu,
+        });
+
+        const $appointment = $('.dx-scheduler-appointment');
+        $appointment.trigger('dxclick');
+
+        this.clock.tick(300);
+
+        let $appointmentItem = $('.dx-list-item');
```
Comment (Raushen) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets.scheduler/common.events.tests.js]: Let's declare the variable before usage

File: packages/devextreme/testing/tests/DevExpress.ui.widgets.scheduler/common.events.tests.js [hunk 4]
```diff
@@ -718,4 +718,44 @@ QUnit.module('Events', {
 
         assert.equal(contentReadyFiresCount, 3, 'contentReadyFiresCount === 3');
     });
+
+    QUnit.test('onAppointmentContextMenu should be triggered when items in the appointment tooltip and appointment collector tooltip is right-clicked (T1181442)', function(assert) {
+        const onAppointmentContextMenu = sinon.spy(({ appointmentElement }) => {
+            assert.ok($(appointmentElement).is($appointmentItem), 'same element');
+        });
```
Comment (Raushen) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets.scheduler/common.events.tests.js:726]: How about:
```
let eventAppointmentElement = null;
const onAppointmentContextMenu = sinon.spy(({ appointmentElement }) => {
   eventAppointmentElement = appointmentElement;
});
...
 assert.ok($(eventAppointmentElement).is($appointmentItem), 'same element');
eventAppointmentElement = null;
...
```

File: packages/devextreme/testing/tests/DevExpress.ui.widgets.scheduler/common.events.tests.js (hunk 3 above)
Comment (markallenramirez) [reply packages/devextreme/testing/tests/DevExpress.ui.widgets.scheduler/common.events.tests.js]: refactored, variable is now declared before usage

File: packages/devextreme/testing/tests/DevExpress.ui.widgets.scheduler/common.events.tests.js (hunk 4 above)
Comment (markallenramirez) [reply packages/devextreme/testing/tests/DevExpress.ui.widgets.scheduler/common.events.tests.js:726]: refactored with some small tweak

**Review by Raushen — DISMISSED** [review]

File: packages/devextreme/js/ui/scheduler/ui.scheduler.js [hunk 5]
```diff
@@ -1352,10 +1353,26 @@ class Scheduler extends Widget {
...
         };
     }
 
+    _createEventArgs(e) {
+        const config = {
+            itemData: e.itemData.appointment,
+            itemElement: e.itemElement,
+            targetedAppointment: e.itemData.targetedAppointment,
+        };
+        return extend({}, this.fire('mapAppointmentFields', config), {
```
Comment (andrewmakarov) [inline packages/devextreme/js/ui/scheduler/ui.scheduler.js:1368]: Can **extend function** be replaced by the **spread operator**?

File: packages/devextreme/testing/tests/DevExpress.ui.widgets.scheduler/common.events.tests.js [hunk 6]
```diff
@@ -718,4 +718,49 @@ QUnit.module('Events', {
...
+            currentDate: new Date(2023, 7, 18),
+            onAppointmentContextMenu,
+        });
+
+        const $appointment = $('.dx-scheduler-appointment');
+        $appointment.trigger('dxclick');
+
+        this.clock.tick(300);
+
+        let $appointmentItem = $('.dx-list-item');
```
Comment (andrewmakarov) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets.scheduler/common.events.tests.js]: All this code is hidden in the createWrapper object. See examples in source code and declaration this object

**Review by andrewmakarov — CHANGES_REQUESTED** [review]

File: packages/devextreme/js/ui/scheduler/ui.scheduler.js (hunk 5 above)
Comment (markallenramirez) [reply packages/devextreme/js/ui/scheduler/ui.scheduler.js:1368]: The [extend function's ability to exclude null fields](https://github.com/DevExpress/DevExtreme/blob/23_2/packages/devextreme/js/core/utils/extend.js#L30) is required here. For instance, [onAppointmentContextMenu](https://js.devexpress.com/Documentation/ApiReference/UI_Components/dxScheduler/Configuration/#onAppointmentContextMenu)'s `model` event property is conditionally available, and so must be defined in the resulting event object only if available.

Technically, the same logic can be done with spread operator (e.g., implement the same if statement in the extend function), but I think it's better to just reuse available utility methods.

File: packages/devextreme/testing/tests/DevExpress.ui.widgets.scheduler/common.events.tests.js (hunk 6 above)
Comment (markallenramirez) [reply packages/devextreme/testing/tests/DevExpress.ui.widgets.scheduler/common.events.tests.js]: refactored

**Review by andrewmakarov — APPROVED** [review]

**Pull request merged**